src/librustc_span/lib.rs

```diff
@@ -836,30 +836,42 @@ pub struct NormalizedPos {
     pub diff: u32,
 }
 
-/// The state of the lazy external source loading mechanism of a `SourceFile`.
-#[derive(PartialEq, Eq, Clone)]
+#[derive(PartialEq, Eq, Clone, Debug)]
 pub enum ExternalSource {
+    /// No external source has to be loaded, since the `SourceFile` represents a local crate.
+    Unneeded,
+    Foreign {
+        kind: ExternalSourceKind,
+        /// This SourceFile's byte-offset within the source_map of its original crate
+        original_start_pos: BytePos,
+        /// The end of this SourceFile within the source_map of its original crate
+        original_end_pos: BytePos,
+    },
+}
+
+/// The state of the lazy external source loading mechanism of a `SourceFile`.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum ExternalSourceKind {
```

Comment:
Separating `ExternalSourceKind` no longer makes sense now when both are enums, could you merge `ExternalSourceKind` back into `ExternalSource`?

Reply:
That would require me to duplicate `original_start_pos` and  `original_end_pos` in each variant, which seems worse.

Reply:
I assumed only `ExternalSourceKind::Present` needs these fields, but that's not the case?

Reply:
We need it for all variants (e.g. https://github.com/Aaron1011/rust/blob/24e7be7186ab084095e5e898251c81a67b2c4026/src/librustc_span/source_map.rs#L340-L343)

`original_start_pos`/`original_end_pos` are a property of the decoded `SourceFile` itself, while `ExternalSourceKind` just describes the current state of our attempt to load the actual source code from disk.

Reply:
I see.